akd/src/directory.rs

```diff
@@ -580,6 +572,145 @@ where
+        params: HistoryParams,
+    ) -> Result<(HistoryProofV2, EpochHash), AkdError> {
+        // The guard will be dropped at the end of the proof generation
+        let _guard = self.cache_lock.read().await;
+
+        let current_azks = self.retrieve_azks().await?;
+        let current_epoch = current_azks.get_latest_epoch();
+        let mut user_data = self.storage.get_user_data(akd_label).await?.states;
+
+        // reverse sort from highest epoch to lowest
+        user_data.sort_by(|a, b| b.epoch.cmp(&a.epoch));
+
+        // apply filters specified by HistoryParams struct
+        user_data = match params {
+            HistoryParams::Complete => user_data,
+            HistoryParams::MostRecent(n) => user_data.into_iter().take(n).collect::<Vec<_>>(),
+        };
+
+        if user_data.is_empty() {
+            let msg = if let Ok(username_str) = std::str::from_utf8(akd_label) {
+                format!("User {username_str}")
+            } else {
+                format!("User {akd_label:?}")
+            };
+            return Err(AkdError::Storage(StorageError::NotFound(msg)));
+        }
+
+        #[cfg(feature = "preload_history")]
+        {
+            let mut lookup_infos = vec![];
+            for ud in user_data.iter() {
+                if let Ok(lo) = self.build_lookup_info(ud).await {
+                    lookup_infos.push(lo);
+                }
+            }
+            current_azks
+                .preload_lookup_nodes(&self.storage, &lookup_infos)
+                .await?;
+        }
+
+        let mut update_proofs = Vec::<UpdateProof>::new();
+        let mut start_version = user_data[0].version;
+        let mut end_version = 0;
+        for user_state in user_data {
+            // Ignore states in storage that are ahead of current directory epoch
+            if user_state.epoch <= current_epoch {
+                let proof = self
+                    .create_single_update_proof(akd_label, &user_state)
+                    .await?;
+                update_proofs.push(proof);
+                start_version = if user_state.version < start_version {
+                    user_state.version
+                } else {
+                    start_version
+                };
+                end_version = if user_state.version > end_version {
+                    user_state.version
+                } else {
+                    end_version
+                };
+            }
+        }
+
+        if start_version == 0 {
+            return Err(AkdError::Directory(DirectoryError::InvalidVersion(
+                "Computed start version for the key history should be non-zero".to_string(),
+            )));
+        }
+
+        let (past_marker_versions, future_marker_versions) =
+            get_marker_versions(start_version, end_version, current_epoch);
+
+        let mut past_marker_vrf_proofs = Vec::<Vec<u8>>::new();
+        let mut existence_of_past_marker_proofs = Vec::<MembershipProof>::new();
```

Comment:
Nit: Any reason we're using the turbofish notation here? Not correctly inferred without it?

Reply:
not sure if I follow -- what is the alternative that you suggest?

Reply:
I think Dillon's question is, can we simplify these lines to:

```
        let mut past_marker_vrf_proofs = vec![];
        let mut existence_of_past_marker_proofs = vec![];
```
and not explicitly specify the types, since presumably they are inferred